src/FlattenNestedRamps.cpp

```diff
@@ -0,0 +1,59 @@
+#include "FlattenNestedRamps.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+namespace {
+
+class FlattenRamps : public IRMutator {
+    using IRMutator::visit;
+
+    Expr visit(const Ramp *op) override {
+        if (op->base.type().is_vector()) {
+            Expr base = mutate(op->base);
+            Expr stride = mutate(op->stride);
+            std::vector<Expr> ramp_elems;
+            for (int ix = 0; ix < op->lanes; ix++) {
+                ramp_elems.push_back(base + ix * stride);
+            }
+
+            return Shuffle::make_concat(ramp_elems);
+        }
+
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const Broadcast *op) override {
+        if (op->value.type().is_vector()) {
+            Expr value = mutate(op->value);
+            std::vector<Expr> broadcast_elems;
+            for (int ix = 0; ix < op->lanes; ix++) {
+                broadcast_elems.push_back(value);
+            }
+
+            return Shuffle::make_concat(broadcast_elems);
```

Comment:
Sorry if I misunderstood your suggestion, but value can be a vector, so I think `std::vector<int>(op->lanes, 0)` won't work. However, value is indeed repeating in the broadcast, so we can use that with something like Shuffle::make({value}, [vector with numbers from 0 to value->lanes] repeated op->lanes times), which might be more efficient than concat.

Reply:
`std::vector<int>(op->lanes, 0)` is a vector of 0s with length op->lanes, i.e. this implements what you are suggesting :)

I think it would also be good to add a `Shuffle::make_broadcast` helper that does this.

Reply:
Right, this will indeed work for scalar value, but not for vector :)

I've added Shuffle::make_broadcast.